### DISCUSSION TITLE
TCP Packet Spec

### DISCUSSION THREAD

**Discussion by rwb196884** [celsworth/lxp-bridge#193] [General]
As described at https://github.com/celsworth/lxp-bridge/wiki/TCP-Packet-Spec

The -heartbeat_ packet seems always to have length 19 -- the 18 byte header plus one zero byte in the 'payload', but I'm seeing _packet length_ values (bytes 4 and 5) of 13.

I'm experiencing packets with `function` (byte 7) value of zero.

What are
```
0xc3 | Read Param
0xc4 | Write Param
```
? (Never seen one.)

I'm seeing my inverter serial number in the 10 bytes starting at index 22 (not 21) in the case of _translated data_. That leaves two bytes unaccounted for.
I'd buess that 20 and 21 

What does 'Address?' mean as the description of _translated data_ payload packet at index zero (index 18 in the whole packet)?

I'm getting zeros at bytes `18 + 4 + 10` and `18 + 4 + 10 + 1` for the _translated data_ length.

**Comment by celsworth** [comment]
Apologies for that page being a bit of a mess, I'm not sure I ever expected anyone to read it ;)

I added a couple of small clarifications around the addresses. There's two per packet as you've seen, one in the TCP header (the first 18 bytes) which always seems to be 1, and one in the translated data part, which is 0 when writing to inverter, and 1 when reading packets from it.

0xc3 and 0xc4 are for reading/writing datalogger parameters, so settings on the wifi stick itself. Some people have had luck setting Wifi AP/password etc with this, my own datalogger doesn't seem to respond to much though.

Your heartbeats being different are concerning, what protocol version do you have in the tcp header? Seems like there might be some changes in newer ones that we have to cope with?

I've never seen a function byte of zero either, that just seems weird 🤔 

Could you post some hexdumps of some full packet samples along with what you think they are? Replace any serial numbers with `0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF` or whatever.

**Comment by rwb196884** [comment]
Hey. Have just dug this out.
I have been unable to guess how to convert 10 bytes into my serial number.

Here is a heartbeat message. My socket receives 19 bytes:
```
0  : A1 // Prefix
1  : 1A // Prefix
2  : 2
3  : 0
4  : D // Packet length 13
5  : 0
6  : 1   // Address is 1
7  : C1 // TCP Function for Heartbeat
8  : 42
9  : 41
10 : 31
11 : 32
12 : 32
13 : 35
14 : 30
15 : 33
16 : 37
17 : redacted
18 : 0
```

(This is in the payload of the TCP messages, not the packet headers.)

**Comment by rwb196884** [comment]
How do you read the _translated data_ starting at byte 18?

**Comment by rwb196884** [comment]
Seems that my inverter is sending the holding registers periodically but not the input registers. Any idea how to get it to send the inputs?
